### ISSUE TITLE
zio-test: `macro has not been expanded` for test with flatMap

### ISSUE THREAD

**Issue by felix-hedenstrom** [zio/zio#8668]
Tested this on `2.0.21` and `2.0.15`. I wrote a test which runs and effect, `flatMap`s the result and does an assertion. The result from trying to compile the test is
```
[error] FooSpec.scala:8:79: macro has not been expanded
[error]   override def spec: Spec[TestEnvironment with Scope, Any] = suite("FooSuite")(
[error]                                                                               ^
[error] one error found
[error] (Test / compileIncremental) Compilation failed
[error] Total time: 28 s, completed Feb 12, 2024, 3:41:55 PM
```

Here is the test
```scala
import zio.{Scope, ZIO}
import zio.test._

object FooSpec extends ZIOSpecDefault {

  val foo = ZIO.succeed(1)

  override def spec: Spec[TestEnvironment with Scope, Any] = suite("FooSuite")(
    test("foo test")(
      foo.flatMap(result => assertTrue(result == 12))
    )
  )
}
```

If I rewrite the body of the test to be 
```scala
for {
  result <- foo
} yield assertTrue(result == 12)
```
it works as expected.

**Comment by hearnadam** [comment]
I don't think this is a bug. The last line in `for` invokes `.map`, as such, the `.flatMap` example given is not equivalent to the `for` example.

FlatMap accepts a function that returns a ZIO.
```scala
def flatMap[R1 <: R, E1 >: E, B](k: A => ZIO[R1, E1, B])(implicit trace: Trace): ZIO[R1, E1, B]
```
[source](https://github.com/zio/zio/blob/series/2.x/core/shared/src/main/scala/zio/ZIO.scala#L628C3-L629)

Map accepts a function that returns anything:
```scala
def map[B](f: A => B)(implicit trace: Trace): ZIO[R, E, B] =
  flatMap(a => ZIO.succeed(f(a)))
```
[source](https://github.com/zio/zio/blob/series/2.x/core/shared/src/main/scala/zio/ZIO.scala#L959-L960)

**Comment by felix-hedenstrom** [comment]
You are right of course. Could the error message be improved?

**Comment by jdegoes** [comment]
/bounty $50 for improving the error message without breaking backward compatibility.

**Comment by vivasvan1** [comment]
/attempt #8668

<div id="algora" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clthhvs2o0017jt0fdq6a36s3/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by varshith257** [comment]
/attempt #8668

<div id="algora" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clthhvs2o0017jt0fdq6a36s3/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by varshith257** [comment]
@hearnadam Can you help me? I am searching for the code where this error `macro has not been expanded` is emitting from. I  can't find in this large codebase which I am trying from last 6 hours to locate the file that emitting this error

**Comment by desavitsky** [comment]
I found earlier, that it is deeply in scala std library, precisely, `RefChecks` class

**Comment by varshith257** [comment]
> I found earlier, that it is deeply in scala std library, precisely, `RefChecks` class

@jdegoes @hearnadam @felix-hedenstrom  Any suggestions on this?

**Comment by hearnadam** [comment]
/attempt #8668

<div id="algora-attempt" />

| [Algora profile](https://console.algora.io/@/hearnadam) | Completed bounties | Tech | Active attempts | Options |
| --- | --- | --- | --- | --- |
| @hearnadam | 4 bounties from 1 project | <div align="center">Scala, Shell, <br />Rust</div> | <div align="center"></div> | [Cancel attempt](https://console.algora.io/api/bounties/clthhvs2o0017jt0fdq6a36s3/cancel-attempt) |

**Comment by noshamexgpt** [comment]
/attempt #8668

Plan: reproduce the flatMap/assertTrue compile failure, identify the narrowest place to surface a clearer diagnostic without changing assertion semantics or public APIs, add focused regression coverage if feasible, and submit a small PR with the exact validation steps.

**Comment by SAVAMforWIN** [comment]
/attempt #8668

Plan: reproduce the macro has not been expanded failure around latMap-based zio-test assertions, isolate whether the break is in the assertion macro expansion order or in the smart-assertion rewrite path, and then submit a narrow regression fix with compiler-focused coverage instead of a broad assertion refactor.

**Comment by zergzorg** [comment]
/attempt #8668

Plan: make a narrow zio-test diagnostic fix for the `flatMap(result => assertTrue(...))` case without changing assertion semantics. I will reproduce the compiler path locally, add focused compile-time/regression coverage if the repository already has a matching test harness, and submit a small PR that turns the misleading `macro has not been expanded` failure into an actionable error for users.
